### PR TITLE
Tasks on SPA can have on_cancel_task #1727

### PR DESCRIPTION
**Pull request by maheshp**
* Added on_cancel attribute to Task model
* Added onCancel widget
* TaskConfigWidget is split into multiple sub_components

### CODE REVIEW

File: server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/helpers/dragula_config.js [hunk 1]
```diff
@@ -22,7 +22,10 @@ define(['dragula'], function (dragula) {
 
     var drake = dragula([elem], {
       revertOnSpill:   true,
-      mirrorContainer: elem
+      mirrorContainer: elem,
+      invalid: function(el, h) {
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/helpers/dragula_config.js]: Can we pass this as an option to `dragulaConfig`? Seems like the implementation is very specific to tasks

File: server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/models/tasks.js [hunk 2]
```diff
@@ -47,6 +59,14 @@ define(['mithril', 'lodash', 'string-plus', './model_mixins', './argument', './r
     this._attributesToJSON = function () {
       throw new Error("Subclass responsibility!");
     };
+
+    this.hasOnCancelTask = function () {
+      return ! _.isNil(this.onCancelTask);
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/models/tasks.js]: Did not know about `isNil`, can you get rid of `isUndefinedOrNull` in `string-plus.js` (or change impl to `isUndefinedOrNull: _.isNil`)

**Comment by ketan** [comment]
Few general observations and peeves I have with the runif options (mostly labelling changes to clarify intent) —

Run if:
* the job has passed (probably even check this by default, since it is the default)
* the job has failed
* always (don't know what's a better label)

Missing lookup for oncancel?

Can we improve the "On Cancel Task" label to "If the job is cancelled when this stage is running", or something similar.

File: server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/helpers/dragula_config.js (hunk 1 above)
Comment (maheshp) [inline server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/helpers/dragula_config.js]: You are right, I should have passed it, will do.

File: server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/models/tasks.js (hunk 2 above)
Comment (maheshp) [inline server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/models/tasks.js]: will get rid of ```isUndefinedOrNull``` for now

**Comment by maheshp** [comment]
> Missing lookup for oncancel?

I know bout it, there are some issues with the UI, will do once Naveen is around

**Comment by maheshp** [comment]
> Run if:
>
> the job has passed (probably even check this by default, since it is the default)
>the job has failed
>always (don't know what's a better label)

`the job has passed` seems a little misleading since the job is in progress. How bout this,


`Run If Job Status` or `Run If Current Job Status`,

- Passed
- Failed
- Any

**Comment by ketan** [comment]
Sounds a lot better, lets go with what you have.

**Pull request merged**